### PR TITLE
File target: Handle UnauthorizedAccessException for multi-process mutexes (e.g. Xamarin and UWP)

### PR DESCRIPTION
**Pull request by 304NotModified**
detect creating Mutex also runtime.

alternative solution for #3048 

See #2742 and #2824 and #2925 and #3059

supersedes #3063

### CODE REVIEW

**Comment by 304NotModified** [comment]
@snakefoot your input please :)

**Comment by 304NotModified** [comment]
> So when people configure KeepFileOpen on limited NetCore-platforms then they will continue to have KeepFileOpen=false. Think this is very confusing, and still prefer to have ConcurrentWrites = false by default on NetCore-platforms, instead of all this guess-magic.

I tried to *not* change the current behavior (the fallback when there is no mutex), but only change the detection, which wasn't good enough.

If the fallback is not good, IMO that's another issue and maybe more for NLog 5.

> Still don't like the testing with exceptions. Also you need to catch a lot more. See also
> 
> `if (ex is SecurityException || ex is UnauthorizedAccessException || ex is NotSupportedException || ex is NotImplementedException || ex is PlatformNotSupportedException)`

will fix that

**Comment by snakefoot** [comment]
Please make `CreateSharableMutex()` return null instead of creating a dummy Mutex-object.

And places where null is not acceptable (Ex. `MutexMultiProcessFileAppender`) then throw exception.

**Comment by 304NotModified** [comment]
> Please make `CreateSharableMutex()` return null instead of creating a dummy Mutex-object.

that's existing code. Just moved around.

**Comment by snakefoot** [comment]
Yes but now you have improve the detection, and the fallback is invalid, and now the fallback is exercised even more.

**Comment by 304NotModified** [comment]
Fair enough, will check that

**Comment by snakefoot** [comment]
Btw. Title of PR should be UWP (and not UPW)

**Comment by 304NotModified** [comment]
WUPS!

**Comment by 304NotModified** [comment]
@snakefoot please review, thanks!

File: src/NLog/Internal/FileAppenders/BaseMutexFileAppender.cs [hunk 1]
```diff
@@ -127,10 +136,15 @@ protected override void Dispose(bool disposing)
         protected Mutex CreateSharableMutex(string mutexNamePrefix)
         {
             if (!PlatformDetector.SupportsSharableMutex)
-                return new Mutex();
+                throw new Exception("Creating Mutex not supported");
```
Comment (snakefoot) [inline src/NLog/Internal/FileAppenders/BaseMutexFileAppender.cs]: Maybe a better Exception-Type. Ex. `System.NotSupportedException`

File: src/NLog/Internal/PlatformDetector.cs [hunk 2]
```diff
@@ -76,19 +85,60 @@ public static bool SupportsSharableMutex
...
+                catch (Exception ex) 
+                {
+                    InternalLogger.Debug(ex, "Failed to create sharable mutex processes");
+                    runTimeSupportsSharableMutex = false;
+                }
+
+                return runTimeSupportsSharableMutex.Value;
+                }
+
+            }
```
Comment (snakefoot) [inline src/NLog/Internal/PlatformDetector.cs]: Please re-format the cs-file, so curly-braces are indented properly.

File: src/NLog/Targets/FileTarget.cs [hunk 3]
```diff
@@ -449,6 +449,12 @@ public bool ConcurrentWrites
             get
             {
 #if SupportsMutex
+
+                if (!PlatformDetector.SupportsSharableMutex)
+                {
+                    return _concurrentWrites ?? false;  // Better user experience for mobile platforms
```
Comment (snakefoot) [inline src/NLog/Targets/FileTarget.cs]: Maybe change into `return _concurrentWrites ?? PlatformDetector.SupportsSharableMutex` ?

File: src/NLog/Targets/FileTarget.cs [hunk 4]
```diff
@@ -1747,10 +1753,30 @@ private bool TryArchiveFile(string fileName, LogEventInfo ev, int upcomingWriteS
...
+                            {
+                                mutexFileAppender.ArchiveMutex.WaitOne();
+                            }
+                        }
                         else if (!KeepFileOpen || ConcurrentWrites)
+                        {
+                            mutexSupportedForArchive = false;
+                        }
+
+                        if (!mutexSupportedForArchive)
```
Comment (snakefoot) [inline src/NLog/Targets/FileTarget.cs]: Maybe just change into:

```
if (archivedAppender is BaseMutexFileAppender mutexFileAppender && mutexFileAppender.ArchiveMutex != null)
{
     mutexFileAppender.ArchiveMutex.WaitOne();
}
else
{
     InternalLogger.Info("FileTarget(Name={0}): Archive mutex not available: {1}", Name, archiveFile);
}
```

**Comment by 304NotModified** [comment]
@snakefoot thanks for your suggestions. Those are applied now. Please let me know if this is ready to merge :)

**Comment by snakefoot** [comment]
Looks better now. Still prefer just having global-mutex disabled by default for NetCore. Better performance and better compatibility.

**Comment by 304NotModified** [comment]
> Looks good now. Still prefer just having global-mutex disabled by default for NetCore. Better performance and better compatibility. Maybe for NLog 5.0 ?

I think that's OK, but we have to double check the pro and cons for it

**Pull request merged**